core/src/main/java/org/mule/runtime/core/internal/streaming/StreamingGhostBuster.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.streaming;
+
+
+import static java.lang.Thread.currentThread;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.slf4j.LoggerFactory.getLogger;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.api.lifecycle.Lifecycle;
+import org.mule.runtime.api.scheduler.Scheduler;
+import org.mule.runtime.api.scheduler.SchedulerConfig;
+import org.mule.runtime.api.scheduler.SchedulerService;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+
+/**
+ * Tracks instances of {@link ManagedCursorProvider} through the {@link #track(ManagedCursorProvider)} method.
+ * This class uses a {@link ReferenceQueue} so that when each of those instances are garbage collected, we can make sure
+ * that {@link ManagedCursorProvider#releaseResources()} is invoked.
+ * <p>
+ * This is useful in cases of long running flows in which cursor providers are open and dereferenced long before the flow ends
+ * (e.g: an &lt;until-successful&gt; which reads a file many times or a &lt;foreach&gt; that executes many queries which result
+ * are never used outside of the scope).
+ * <p>
+ * Notice that this <b>DOES NOT</b> replaces the cleanup and tracking job that is performed in {@link CursorManager}. That
+ * still remains the main cleanup mechanism as we cannot rely on garbage collection for releasing streaming resources. This is
+ * only a mitigation for cases like described above.
+ *
+ * @since 4.2.0
+ */
+public class StreamingGhostBuster implements Lifecycle {
+
+  private static final long POLL_INTERVAL = SECONDS.toMillis(5);
+  private static final Logger LOGGER = getLogger(StreamingGhostBuster.class);
+
+  private final ReferenceQueue<ManagedCursorProvider> referenceQueue = new ReferenceQueue<>();
+  private volatile boolean stopped = false;
+  private Future taskHandle;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private Scheduler scheduler;
+
+  @Override
+  public void initialise() throws InitialisationException {
+    scheduler = schedulerService.customScheduler(SchedulerConfig.config()
```

Comment:
This could be a periodic task on an IO thread...

Reply:
it could. But should it? The QueueReference has a blocking method. Isn't it better to use that to have one dedicated thread leveraging that?

Reply:
don't have an opinion on this really. Maybe in the future we should do some tests around this.